pkg/operator/status.go

```diff
@@ -97,57 +87,84 @@ func isDeploymentStatusComplete(deploy *appsv1.Deployment) bool {
 		deploy.Status.ObservedGeneration >= deploy.Generation
 }
 
-func (c *serviceCAOperator) syncStatus(operatorConfigCopy *operatorv1.ServiceCA, deployments []string) (bool, error) {
-	ready := 0
-	existingDeploymentsAndReplicas := 0
-	for _, dep := range deployments {
-		existing, err := c.appsv1Client.Deployments(operatorclient.TargetNamespace).Get(dep, metav1.GetOptions{})
-		if err != nil {
-			if apierrors.IsNotFound(err) {
-				statusMsg := fmt.Sprintf("Deployment %s does not exist", dep)
-				c.setProgressingStatus(operatorConfigCopy, statusMsg)
-				return false, nil
+func (c *serviceCAOperator) syncStatus(operatorConfigCopy *operatorv1.ServiceCA, existingDeps *appsv1.DeploymentList, targetDeps []string) {
+	versionReady := true
+	deploymentsComplete := true
+	statusMsg := ""
+	for _,  target := range targetDeps {
```

Comment:
I'm not quite sure this does what I thought it was going to do, the original might have been a bit better. How about this:

Have 4 strings - `notFound`, `noReplicas`, `notYetComplete` or similar, they correspond to the original conditions (if you're bringing any new here, feel free to add, I did not check fully).

In each of the conditions, if it applies for the deployment, do `<thisConditionStringName> += fmt.Sprintf("\n%s", deploymentName)` and instead of the original `return false, nil`, just do `continue`

After the loop `for _, dep := range deployments {` finishes (keep `dep` here, btw, using shortcuts in loops is fine IMHO), check the string vars you had and do (for example):
```go
progressingMsg := ""
if len(notFound) != 0 {
    progressingMsg += fmt.Sprintf("The following deployments were not yet created:\n%s", notFound)
}
```
and eventually set the `Progressing` and `Available` statuses accordingly, with the progressing status having the long message, available can just be more brief.

Please do reach out to me if any of this is not clear :wink:

Reply:
yes, I am in the process of rechecking the logic, etc.. still wip.  it's clear, thanks : ) also I agree the syncStatus needs rework

Reply:
@stlaz , I've cleaned  up syncStatus (had some extra booleans there) : )  now looking at your suggestion, maybe I'll revert back to a cleaner version of what I had b4. 
> Have 4 strings - `notFound`, `noReplicas`, `notYetComplete` or similar, they correspond to the original conditions (if you're bringing any new here, feel free to add, I did not check fully).
> 
> In each of the conditions, if it applies for the deployment, do `<thisConditionStringName> += fmt.Sprintf("\n%s", deploymentName)` and instead of the original `return false, nil`, just do `continue`

Reply:
@stlaz @mrogers950 I think syncStatus() makes sense the way it is atm.  We are following just 3 deployments, they are all deployed w/ the same operator image, so unless somehow the deployment yaml was modified (b4 install? since it's managed via CVO) the notFound will not happen (unless something is horribly wrong cluster-wide).  For this component once Progressing = true during an upgrade, a status for 1 of the deployments will suffice as overall CO status msg (since they all use same image, managed via CVO), and if progressing=true because a deployment was manually deleted, the correct status will be reported for that deployment. 

I've gathered `oc get serviceca -o yaml` during both an upgrade and when I manually delete the deployments in openshift-service-ca.  The status/reasons/messages all are as expected.

Reply:
https://github.com/openshift/service-ca-operator/pull/38 makes it possible for the deployments to get redeployed, I suspect `notFound` would still be possible.

Would deployments each be always in the same state, even during upgrade/start? Could replicas of one deployment be slower of the other? I guess they could, but correct me if I'm wrong, I am not familiar with the corner cases of service-ca operator :slightly_smiling_face: 

In both the first case and the second one, though, some replicas may not be found, some may be late and having the information about each the situation for both IMHO makes sense.

Also, to simplify the code a bit, consider moving the `targetDeploymentNames` slice into a set (https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets/string.go#L30) and use `x.Has(dep)` instead.

Reply:
> Also, to simplify the code a bit, consider moving the `targetDeploymentNames` slice into a set (https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets/string.go#L30) and use `x.Has(dep)` instead.
thanks :)

Reply:
_much better_